### PR TITLE
SFX-148: Incorporate interaction testing against Search component

### PR DESCRIPTION
**Pull request by JoeyDeol**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -201,6 +222,10 @@ workflows:
       - unit_test:
           name: unit_test_search_box
           <<: *requires_build_final
+
+      - unit_test_interaction:
+          name: unit_test_interaction_search_box
+          <<: *requires_build_final
       - unit_test_final:
```
Comment (jmbtutor) [inline .circleci/config.yml]: Add a blank line above to separate interaction tests from the final collector. That, or remove the blank line separating the unit tests from the interaction tests.

File: .circleci/config.yml [hunk 2]
```diff
@@ -138,6 +138,27 @@ jobs:
             cd packages/*/"@sfx/${package}"
             yarn test
 
+  ### unit_test_interaction: unit_test_interaction_{package}
+  ### Runs the interaction tests for {package}.
+  unit_test_interaction:
```
Comment (jmbtutor) [inline .circleci/config.yml]: I don't think interaction tests, at least the way we've written them, count as unit tests. I'd call this `interaction_test`, then change `unit_test_final` to `test_final` or something like that.

File: interaction.karma.conf.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+const webpackConfig = require('./webpack.config');
+const path = require('path');
+
+delete webpackConfig.entry;
+
+module.exports = (config) => {
+  config.set({
```
Comment (jmbtutor) [inline interaction.karma.conf.js]: This is basically the same as the regular config, so I'd export a function that creates a default config, which a config file further down can import and modify as needed.

File: interaction.karma.conf.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    browsers: ['ChromeHeadless'],
+    frameworks: ['mocha', 'sinon-chai'],
+    reporters: ['spec', 'coverage-istanbul'],
+    files: [
+      'test/*.ts',
+      'test/interaction/*.test.ts'
+    ],
+    preprocessors: {
+      'test/*.ts': ['webpack'],
+      'test/interaction/*.test.ts': ['webpack']
```
Comment (jmbtutor) [inline interaction.karma.conf.js]: I know this was in the other one, but I'm pretty sure this is redundant. The first one should cover it. Double-check, though.

File: karma.conf.js [hunk 5]
```diff
@@ -10,11 +10,11 @@ module.exports = (config) => {
     reporters: ['spec', 'coverage-istanbul'],
     files: [
       'test/*.ts',
-      'test/**/*.test.ts'
+      'test/unit/*.test.ts'
```
Comment (jmbtutor) [inline karma.conf.js]: I think this should be `'test/unit/**/*.test.ts'`.

File: interaction.karma.conf.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+delete webpackConfig.entry;
+
+module.exports = (config) => {
+  config.set({
+    browsers: ['ChromeHeadless'],
+    frameworks: ['mocha', 'sinon-chai'],
+    reporters: ['spec', 'coverage-istanbul'],
+    files: [
+      'test/*.ts',
+      'test/interaction/*.test.ts'
```
Comment (jmbtutor) [inline interaction.karma.conf.js]: I think this should be `'test/interaction/**/*.test.ts'`.

File: packages/web-components/@sfx/search-box/CHANGELOG.md [hunk 7]
```diff
@@ -6,5 +6,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 ### Added
+- SFX-148: Added interaction tests for the SearchBox component.
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/CHANGELOG.md]: I would put this below SFX-151. Adding the interaction tests before the component it's testing doesn't make sense.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    });
+  });
+
+  it('should update the searchbox value property when an input event is dispatched', () => {
+    container.appendChild(searchbox);
+
+    return waitForUpdateComplete(searchbox).then(() => {
+      const searchboxInput = searchbox.querySelector('input');
+      searchboxInput.value = "Search Term";
+      const inputEvent = new Event('input', { 'bubbles': true });
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: I would switch this line and the line above to group `const`s together.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+      const searchboxInput = searchboxNode.querySelector('input');
+
+      searchbox.value = searchboxInput.value = "Search Term";
+
+      return waitForUpdateComplete(searchbox);
+    }).then(() => {
+      const searchboxNode = document.querySelector('sfx-search-box');
+      const searchboxButtons = searchboxNode.querySelectorAll('button');
+      const searchboxClearButton = searchboxButtons[0];
+      searchboxClearButton.click();
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: This looks like an action, so I think we should add a blank line above to separate it from the setup.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+    return waitForUpdateComplete(searchbox).then(() => {
+      const searchboxNode = document.querySelector('sfx-search-box');
+      const searchboxInput = searchboxNode.querySelector('input');
+
+      searchbox.value = searchboxInput.value = "Search Term";
+
+      return waitForUpdateComplete(searchbox);
+    }).then(() => {
+      const searchboxNode = document.querySelector('sfx-search-box');
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: This is not necessary. You can use `searchbox` directly.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+  });
+
+  it('should clear the searchbox when the clear button is clicked', () => {
+    searchbox.clearButton = true;
+    searchbox.searchButton = true;
+
+    container.appendChild(searchbox);
+
+    return waitForUpdateComplete(searchbox).then(() => {
+      const searchboxNode = document.querySelector('sfx-search-box');
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: This is not necessary. You can use `searchbox` directly.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+      return waitForUpdateComplete(searchbox);
+    }).then(() => {
+      const searchboxNode = document.querySelector('sfx-search-box');
+      const searchboxButtons = searchboxNode.querySelectorAll('button');
+      const searchboxClearButton = searchboxButtons[0];
+      searchboxClearButton.click();
+
+      return waitForUpdateComplete(searchbox);
+    }).then(() => {
+      const searchboxNode = document.querySelector('sfx-search-box');
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: This is not necessary. You can use `searchbox` directly.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+      const searchboxInput = searchboxNode.querySelector('input');
+
+      expect(searchbox.value).to.equal('');
+      expect(searchboxInput.value).to.equal('');
+    });
+  });
+
+  it('should dispatch a search event when the search button is clicked', () => {
+    searchbox.clearButton = true;
+    searchbox.searchButton = true;
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: When possible, try to keep the following order: `const`, `let`, other assignments. Though I acknowledge for the `let` below that doing so is not possible.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    let eventListenerResolve;
+    const eventListenerPromise = new Promise((resolve) => eventListenerResolve = resolve);
+
+    container.appendChild(searchbox);
+
+    const searchboxNode = container.querySelector('sfx-search-box');
+
+    searchboxNode.addEventListener(SEARCHBOX_EVENT.SEARCH_REQUEST, (e: any) => {
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: Is the `: any` necessary here?

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+  it('should dispatch a search event when the search button is clicked', () => {
+    searchbox.clearButton = true;
+    searchbox.searchButton = true;
+    let eventListenerResolve;
+    const eventListenerPromise = new Promise((resolve) => eventListenerResolve = resolve);
+
+    container.appendChild(searchbox);
+
+    const searchboxNode = container.querySelector('sfx-search-box');
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: Use `searchbox` directly.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+    searchboxNode.addEventListener(SEARCHBOX_EVENT.SEARCH_REQUEST, (e: any) => {
+      eventListenerResolve(e);
+    });
+
+    return waitForUpdateComplete(searchbox).then(() => {
+      const searchboxNode = document.querySelector('sfx-search-box');
+      const searchboxInput = searchboxNode.querySelector('input');
+
+      searchbox.value = searchboxInput.value = "Search Term";
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]:
```suggestion
      searchbox.value = searchboxInput.value = 'Search Term';
```

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    container.appendChild(searchbox);
+
+    const searchboxNode = container.querySelector('sfx-search-box');
+
+    searchboxNode.addEventListener(SEARCHBOX_EVENT.SEARCH_REQUEST, (e: any) => {
+      eventListenerResolve(e);
+    });
+
+    return waitForUpdateComplete(searchbox).then(() => {
+      const searchboxNode = document.querySelector('sfx-search-box');
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: Use `searchbox` directly.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+      return waitForUpdateComplete(searchbox);
+    }).then(() => {
+      const searchboxButtons = searchbox.querySelectorAll('button');
+      const searchboxSearchButton = searchboxButtons[1];
+
+      searchboxSearchButton.click();
+
+      return waitForUpdateComplete(searchbox);
+    }).then(() => {
+
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: This blank line is not necessary.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+      return eventListenerPromise;
+    }).then((e: any) => {
+
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: This blank line is not necessary.

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+      const searchboxButtons = searchbox.querySelectorAll('button');
+      const searchboxSearchButton = searchboxButtons[1];
+
+      searchboxSearchButton.click();
+
+      return waitForUpdateComplete(searchbox);
+    }).then(() => {
+
+      return eventListenerPromise;
+    }).then((e: any) => {
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: Is this type annotation necessary?

File: test-tools/utils.ts [hunk 21]
```diff
@@ -8,3 +8,7 @@ export const chai = global.chai || chaiImport;
 export const expect = chai.expect;
 export const stub = sinon.stub;
 export const spy = sinon.spy;
+
+export function waitForUpdateComplete(node) {
```
Comment (jmbtutor) [inline test-tools/utils.ts]: Add type annotations:

```suggestion
export function waitForUpdateComplete(node: LitElement): Promise<boolean> {
```

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+import SearchBox from '../../src/search-box';
+import { TemplateResult, html } from 'lit-element';
+
+describe('SearchBox Component Interaction Tests', () => {
+  let searchbox;
+  let container;
+
+  before(() => {
+    container = document.createElement('div');
+    container.id = "interaction-container";
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]:
```suggestion
    container.id = 'interaction-container';
```

File: packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts [hunk 23]
```diff
@@ -0,0 +1,120 @@
+import { SEARCHBOX_EVENT } from '../../src/events';
+import { expect, stub, waitForUpdateComplete } from '../utils';
+import SearchBox from '../../src/search-box';
+import { TemplateResult, html } from 'lit-element';
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/interaction/search-box.test.ts]: We've decided that non-local imports should be specified first when possible. For these imports, I would order it as 4, 2, 3, 1 (non-local, test utils, class under test, other supporting modules).

File: packages/web-components/@sfx/search-box/test/unit/search-box.test.ts [hunk 24]
```diff
@@ -1,5 +1,5 @@
 import { SEARCHBOX_EVENT } from '../../src/events';
-import { expect, stub } from '../utils';
+import { expect, stub, waitForUpdateComplete } from '../utils';
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/unit/search-box.test.ts]: This isn't used. I don't think this file should have any changes.

File: interaction.karma.conf.js (hunk 3 above)
Comment (JoeyDeol) [reply interaction.karma.conf.js]: Created a file within the test tools directory to contain the new function.

File: .circleci/config.yml (hunk 2 above)
Comment (JoeyDeol) [reply .circleci/config.yml]: Updated.

File: .circleci/config.yml (hunk 1 above)
Comment (JoeyDeol) [reply .circleci/config.yml]: Updated.

File: .circleci/config.yml (hunk 2 above)
Comment (JoeyDeol) [reply .circleci/config.yml]: Updated.

**Pull request merged**